core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java

```diff
@@ -91,16 +97,36 @@ protected Selectable doAccept(HandlerEvent<StopStartDownscaleDecommissionViaCMRe
             HostGroup hostGroup = hostGroupService.getByClusterIdAndName(cluster.getId(), request.getHostGroupName())
                     .orElseThrow(NotFoundException.notFound("hostgroup", request.getHostGroupName()));
 
+            List<InstanceMetaData> instancesWithServicesNotRunning = cloudInstanceIdToInstanceMetaDataConverter.getNotDeletedAndNotZombieInstances(stack,
+                    request.getHostGroupName(),
+                    request.getRunningInstancesWithServicesNotRunning().stream()
+                            .map(CloudInstance::getInstanceId)
+                            .collect(Collectors.toSet()));
+
             Map<String, InstanceMetaData> hostsToRemove = clusterDecomissionService.collectHostsToRemove(hostGroup, hostNames);
             List<String> missingHostsInCm = Collections.emptyList();
+            Map<String, InstanceMetaData> finalHostsToRemove = hostsToRemove;
             if (hostNames.size() != hostsToRemove.size()) {
                 missingHostsInCm = hostNames.stream()
-                        .filter(h -> !hostsToRemove.containsKey(h))
+                        .filter(h -> !finalHostsToRemove.containsKey(h))
                         .collect(Collectors.toList());
                 LOGGER.info("Found fewer instances in CM to decommission, as compared to initial ask. foundCount={}, initialCount={}, missingHostsInCm={}",
                         hostsToRemove.size(), hostNames.size(), missingHostsInCm);
             }
 
+            List<InstanceMetaData> additionalInstancesWithServicesNotRunningToDecommission = instancesWithServicesNotRunning
+                    .stream().filter(i -> !finalHostsToRemove.containsKey(i.getDiscoveryFQDN())).collect(Collectors.toList());
+
+            if (!additionalInstancesWithServicesNotRunningToDecommission.isEmpty()) {
+                LOGGER.info("Including running instances with services not running as part of hostsToDecommission");
+                int instancesToBeDecommissionedLimit = hostsToRemove.size() - additionalInstancesWithServicesNotRunningToDecommission.size();
+                List<InstanceMetaData> finalInstanceMetadataToRemove = hostsToRemove.values().stream().limit(instancesToBeDecommissionedLimit)
+                        .collect(Collectors.toList());
+                finalInstanceMetadataToRemove.addAll(additionalInstancesWithServicesNotRunningToDecommission);
+                // TODO: is it a good idea to edit what is directly returned by CM APIs?
```

Comment:
This is tricky. The API specified a list of nodes to decommission. I don't think that set of nodes should be ignored (unless they are problematic for some other reason - e.g. deleted / missing, etc). i.e. just because we found nodes which need to be repaired does not mean we should skip nodes which were specified here. Typically, YARN will not know about nodes in bad state (assuming NM not running), so the decommission recommendation continnues to be valid - the specified nodes are not required per current capacity demands.

More on this elsewhere, which talks about enhancing the downscale API.